app/controllers/sign_up/completions_controller.rb

```diff
@@ -0,0 +1,28 @@
+module SignUp
+  class CompletionsController < ApplicationController
+    def show
+      if user_fully_authenticated? && !session[:sp].nil?
```

Comment:
Yes. If it's an empty hash, I consider that empty, just like an empty array is empty. This method relies on `session[:sp]` having stuff inside the hash, so if there's nothing inside the hash, we want to abort. If a service provider is indeed present, but our hash is empty, that's a bigger issue that would be detected by our SAML specs.